my-sms/db/migrate/20210208041633_create_titles.rb

```diff
@@ -0,0 +1,70 @@
+class CreateTitles < ActiveRecord::Migration
+  def up
+    available_titles = %w[Mr Mrs Miss Ms Master]
+
+    # do not perform migration if any unexpected titles found in Students
+    # may be more efficient to do via SQL query?
+    Student.find_in_batches do |batch|
+      batch.uniq{|s| s[:title]}.each do |s|
+        if available_titles.exclude? s.title
+          raise "Unexpected Student title #{s.title} in "\
+          "Student id #{s.id}. Migration expects titles "\
+          "in Students column to be one of #{available_titles}."
+        end
+      end
+    end
```

Comment:
Why not create a title entry for each unique title found?

Reply:
My thinking was that it would be better to be more explicit with what titles are available, since there is a small number.
Presumably the set of titles is a business decision since there is no documented functionality for users to define a custom title. This way that set of titles will be the same no matter where the migration is run.

Reply:
Yea you're right, in a real world scenario there would probably be a list of valid titles given to you in the ticket. We probably don't want to raise an exception here. There are a few quick solutions I can think of to do:

- Don't raise an error but rather just log the titles and student ids that failed. That way we can create a script (rake task) in future to go through and fix those ones that failed
- Raise an error, but rollback any changes made so we're not left in an intermediate state
- Keep the old title in an existing column for a little while so that post migration we can easily see what titles still need to be migrated over without the end user noticing a change

Reply:
I ended up changing this part to just be a warning in 453cd24
I now just add unexpected titles to the table as you suggested.